elastic/kibana
# [Reporting] CSV export: add warning for unexpected row count

## Summary

Closes https://github.com/elastic/kibana/issues/136445

When running a CSV export job, backend code in the reporting plugin pulls data from Elasticsearch through a series of synchronous search calls. The initial response has a field to inform the caller how many documents there are total in the search result (`hits.total`). This PR uses the given total and adds a warning to the report job if the number of generated CSV rows isn't matching the total.

#### Other changes
1. Promote the log level of the scan and scroll requests from `debug` to `info` 
2. Wrap the scan and scroll calls in try/catch, so if there is an error thrown from the calls it will be added to the report job data. 
    - Currently the scan and scroll calls do not emit an error within 2 minutes, even if the search response is throttled by 30 seconds. See the FIXME comment added, which is for later investigation.

## Testing the search timeout issue
1. Use a [debugging proxy](https://github.com/tsullivan/simple-node-proxy) in-between Kibana and Elasticsearch
   1. Start Elasticsearch locally running on port 9200.
   2. Set `elasticsearch.hosts: ['http://localhost:9202']` in config/kibana.yml
   3. Run node-simple-proxy with the command:
       `npm start -- --timeoutTime=30000 --listenPort=9202 --targetUrl=http://localhost:9200 --noSsl`
   4. Once Kibana is running, attempt a CSV export
2. Notice the job takes 2 minutes to fail, and it fails with a general queue timeout error message.
   1. Further investigation is needed in the data search plugin to understand why an error is not triggered for a search request timeout after 30 seconds. 

### Checklist

Delete any items that are not applicable to this PR.

 - [x] Ensure we ask Elasticsearch for an exact number of total hits in the first response.
    - [X] The search request was already sent with the `"track_total_hits": true` param in the body

## Release note
Added a new type of warning in CSV reports, where an error message will appear on the job if fewer CSV rows were generated than expected.
